File: backend/orion/view_managers/interactive/search_manager/search_data_model/leak/search_leak_callback_model.py
from typing import List, Optional, Union
from pydantic import BaseModel, HttpUrl
from datetime import datetime

class suggestion(BaseModel):
    text: str
    offset: int
    length: int
    options: List[dict]

class result_item(BaseModel):
    m_title: str
    m_url: Optional[HttpUrl] = None
    m_base_url: Optional[HttpUrl] = None
    m_content: str
    m_important_content: str
    m_network: Optional[str] = None
    m_content_type: Union[str, List[str]]
    m_weblink: List[HttpUrl] = []
    m_dumplink: List[HttpUrl] = []
    m_name: Optional[str] = None
    m_email_addresses: List[str] = []
    m_industry: Optional[str] = None
    m_phone_numbers: List[str] = []
    m_addresses: List[str] = []
    m_social_media_profiles: List[str] = []
    m_websites: List[str] = []
    m_company_name: Optional[str] = None
    m_logo_or_images: List[str] = []
    m_leak_date: Optional[datetime] = None
    m_data_size: Optional[str] = None
    m_country_name: Optional[str] = None
    m_revenue: Optional[str] = None
    m_hash: str
    m_update_date: datetime
    m_contact_link: Optional[HttpUrl] = None
    m_creation_date: datetime

class search_leak_callback_model(BaseModel):
    Result: List[result_item]
    Suggestions: List[suggestion]
    Page_Count: float

File: client/src/app/shared/model/intel-results/leak/search_leak_callback_model.ts
export class SuggestionOption {
  text: string = "";
  score: number = 0;
  freq: number = 0;

  constructor(init?: Partial<SuggestionOption>) {
    Object.assign(this, init);
  }
}

export class Suggestion {
  text: string = "";
  offset: number = 0;
  length: number = 0;
  options: SuggestionOption[] = [];

  constructor(init?: Partial<Suggestion>) {
    if (init) {
      this.text = init.text || "";
      this.offset = init.offset || 0;
      this.length = init.length || 0;
      this.options = init.options?.map(opt => new SuggestionOption(opt)) || [];
    }
  }
}

export class ResultItem {
  m_title: string = "";
  m_url?: string;
  m_base_url?: string;
  m_content: string = "";
  m_important_content: string = "";
  m_network?: string;
  m_content_type: string[] = [];
  m_weblink: string[] = [];
  m_dumplink: string[] = [];
  m_email_addresses: string[] = [];
  m_phone_numbers: string[] = [];
  m_social_media_profiles: string[] = [];
  m_websites: string[] = [];
  m_company_name?: string | null;
  m_logo_or_images: string[] = [];
  m_leak_date?: string | null;
  m_data_size?: string | null;
  m_country_name?: string | null;
  m_revenue?: string | null;
  m_update_date: string = new Date().toISOString();
  m_hash: string = "";
  m_creation_date: string = new Date().toISOString();
  m_contact_link?: string;

  constructor(init?: Partial<ResultItem>) {
    Object.assign(this, init);
  }

  toJSON() {
    return JSON.stringify(this);
  }
}

export class SearchLeakCallbackModel {
  Result: ResultItem[] = [];
  Page_Count: number = 0;
  Suggestions: Suggestion[] = [];

  constructor(init?: Partial<SearchLeakCallbackModel>) {
    if (init) {
      this.Result = init.Result?.map(r => new ResultItem(r)) || [];
      this.Suggestions = init.Suggestions?.map(s => new Suggestion(s)) || [];
      this.Page_Count = init.Page_Count ?? 0;
    }
  }

  toJSON() {
    return JSON.stringify(this);
  }
}
